### PR TITLE
Add support for OTEL tracing

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

`crictl` now features 3 new CLI parameters:

- `--enable-tracing`: Enable OpenTelemetry tracing. (default: false)
- `--tracing-endpoint`: Address to which the gRPC tracing collector will send spans to. (default: "0.0.0.0:4317")
- `--tracing-sampling-rate-per-million`: Number of samples to collect per million OpenTelemetry spans. Set to 1000000 or -1 to always sample. (default: -1)

The tracer provider will be created on startup and the `Shutdown()` invocation will ensure that all spans are processed before exiting the binary.

The `hack/tracing` directory contains scripts for local testing:

```
> ./hack/tracing/start
…
Everything is ready, open http://localhost:16686 to access jaeger
```

When now running `crictl` with `--enable-tracing`:

```
> sudo ./build/bin/linux/amd64/crictl --enable-tracing ps
```

Then jaeger should show collected traces and spans for the 3 RPCs `ListContainers`, `ImageFsInfo` as well as `Version`:

screenshot

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes-sigs/cri-tools/issues/1317

#### Special notes for your reviewer:
None
#### Does this PR introduce a user-facing change?

```release-note
Added support for OTEL tracing via the new CLI flags `--enable-tracing`, `--tracing-endpoint` and `--tracing-sampling-rate-per-million`.
```

### CODE REVIEW

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/hold

hold for other maintainers to look at. 

Everything looks good, thank you for adding the `hack/tracing`. I think what can be useful as an additional improvements:

1. Add a start span/stop span operations on individual crictl commands so commands will group all the calls they are making and we will also see failed commands - when they failed with arguments validation.
2. I wonder if we may need to hide any sensitive information off the arguments passed. Since it is an optional feature, as-is OK, but maybe later we can add an option to filter out sensitive information
3. We may want to add a computer name as part of a span so we can group spans by it. This way we can see from jaeger who called crictl command

**Comment by saschagrunert** [comment]
> 1. Add a start span/stop span operations on individual crictl commands so commands will group all the calls they are making and we will also see failed commands - when they failed with arguments validation.
> 2. I wonder if we may need to hide any sensitive information off the arguments passed. Since it is an optional feature, as-is OK, but maybe later we can add an option to filter out sensitive information
> 3. We may want to add a computer name as part of a span so we can group spans by it. This way we can see from jaeger who called crictl command

Thanks! I'll follow-up on those :+1:

**Review by feiskyer — APPROVED** [review]
LGTM. Thanks adding this!

**Comment by feiskyer** [comment]
@saschagrunert could you rebase?

**Comment by saschagrunert** [comment]
@feiskyer done :)

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

File: cmd/crictl/main.go
Comment (kwilczynski) [inline cmd/crictl/main.go:257]: @saschagrunert, would it make sense to make this use `127.0.0.1` here? Especially since we are trying to fix such things in CRI-O, for example.

File: hack/tracing/env
Comment (kwilczynski) [inline hack/tracing/env:3]:
```suggestion
# Copyright 2024 The Kubernetes Authors.
```

File: pkg/tracing/tracing.go
Comment (kwilczynski) [inline pkg/tracing/tracing.go:55]: Nit.

```suggestion
	// Only emit spans when the kubelet sends a request with a sampled trace.
```

File: pkg/tracing/tracing.go
Comment (kwilczynski) [inline pkg/tracing/tracing.go:58]: Nit.

```suggestion
	// Or, emit spans for a fraction of transactions.
```

Comment (saschagrunert) [reply pkg/tracing/tracing.go]: Fixed :+1:

File: pkg/tracing/tracing.go
Comment (saschagrunert) [reply pkg/tracing/tracing.go]: Fixed :+1:

File: hack/tracing/env
Comment (saschagrunert) [reply hack/tracing/env]: Fixed :+1:

File: cmd/crictl/main.go
Comment (saschagrunert) [reply cmd/crictl/main.go]: Fixed :+1:

File: hack/tracing/env
Comment (saschagrunert) [reply hack/tracing/env]: Fixed :+1:

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

**Comment by saschagrunert** [comment]
@SergeyKanzhelev can we lift the hold?

**Comment by saschagrunert** [comment]
@SergeyKanzhelev can we lift the hold?

**Comment by SergeyKanzhelev** [comment]
/unhold

sorry for delay

/lgtm

**Pull request merged**
